### PR TITLE
Added packed sequence and sequence parallel to qwen2vl

### PR DESCRIPTION
**Pull request by chrjxj**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

This PR follow neva implmentation and add packed sequence and sequence parallel to qwen2vl

**Collection**:  `qwen2vl`

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- usage example

```bash
cd /path/to/NeMo

torchrun --nproc_per_node=2 scripts/vlm/qwen2vl_finetune.py \
    --devices=2 --tp_size=2 --pp_size=1 \
    --enable_sp --use_packed_sequence \
    --data_type=qwen2vl \
    --lr 0.00005 \
    --max_sequence_length 8192 \
    --gbs 8 --mbs 2 --max_steps 300 \
    --data_path="/projects/dataset/train.json" \
    --image_folder="/projects/dataset/images" 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

@yaoyu-33 

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/vlm/qwen2vl/data/preloaded.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/data/preloaded.py:1]: no need to update

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 2]
```diff
@@ -92,10 +101,11 @@ def qwen2vl_forward_step(model, batch) -> torch.Tensor:
         "video_grid_thw": batch.get("video_grid_thw", None),
         "loss_mask": batch.get("loss_mask", None),
         "labels": batch.get("labels", None),
+        "packed_seq_params": batch.get("packed_seq_params", None),
     }
 
-    if 'cu_seqlens' in batch:
-        forward_args['packed_seq_params'] = get_packed_seq_params(batch)
+    # if 'cu_seqlens' in batch:
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:107]: remove

File: nemo/collections/vlm/qwen2vl/model/base.py [hunk 3]
```diff
@@ -612,8 +651,70 @@ def forward(
...
+
+                        _position_ids, _ = self.get_rope_index(_input_ids, _image_grid_thw, _video_grid_thw, None)
+                        packed_position_ids.append(_position_ids)
+
+                    position_ids = torch.cat(packed_position_ids, dim=2)
+                else:
+                    raise ValueError("Got unexpected cu_seqlens_kv value. {cu_seqlens_kv}")
+            else:
+                # relugar input data
+                position_ids, _ = self.get_rope_index(input_ids, image_grid_thw, video_grid_thw, attention_mask)
```
Comment (yaoyu-33) [inline nemo/collections/vlm/qwen2vl/model/base.py:704]: where is the position ids got used again?

File: nemo/collections/vlm/qwen2vl/data/preloaded.py [hunk 4]
```diff
@@ -808,3 +816,67 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
                 consumed_samples=consumed_samples,
                 consistency_check=False,
             )
+
+
+def convert_to_packed(
```
[the remark this answers is not in the document]
Comment (chrjxj) [reply nemo/collections/vlm/qwen2vl/data/preloaded.py:821]: 1) the target length of input tokens (with media tokens) is calculated differently.   2) here convert_to_packed() doesn't need to calc and return position_ids

File: nemo/collections/vlm/qwen2vl/model/base.py (hunk 3 above)
Comment (chrjxj) [reply nemo/collections/vlm/qwen2vl/model/base.py:704]: self.language_model(),  aka GPTModel's forward() , will take `position_ids` as input param, and use it to calc `rotary_pos_emb`. See:
https://github.com/NVIDIA/Megatron-LM/blob/8416bff56a06771834be00327e1523aff4b20f88/megatron/core/models/gpt/gpt_model.py#L302 
and then
https://github.com/NVIDIA/Megatron-LM/blob/8416bff56a06771834be00327e1523aff4b20f88/megatron/core/models/common/embeddings/rotary_pos_embedding.py#L276

File: nemo/collections/vlm/qwen2vl/data/preloaded.py (hunk 1 above)
Comment (chrjxj) [reply nemo/collections/vlm/qwen2vl/data/preloaded.py]: done

File: nemo/collections/vlm/qwen2vl/model/base.py (hunk 2 above)
Comment (chrjxj) [reply nemo/collections/vlm/qwen2vl/model/base.py]: fixed.

**Review by yaoyu-33 — DISMISSED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
